[SNOW-851220] Integration tests of snowpark function commands

sfc-gh-pjob opened the pull request:
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added new automated tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the section below.

### Changes description
* Fixed `snow snowpark function update` command execution.
* Added integration test for all commands from `snowpark function` group.
* Added some test utils in `tests_integration/testing_utils` directory.
* Changed temporary integration test database prefix from `db_` to `snowcli_integration_tests_db_` to make our trash be easier to identify.
* Fixed imports in `test_sql.py` and `test_warehouse.py`.
* Added `tests_integration` directory to the list of exclusions for `trailing-whitespace` hook.

sfc-gh-astus commented on tests_integration/conftest.py:
```diff
@@ -37,6 +37,12 @@ def invoke(self, *a, **kw):
+            **kwargs,
+        )
+
+    def invoke_with_config_and_integration_connection(self, *args, **kwargs):
```
I think we will only use this method, so maybe rename this one to `invoke` and remove `invoke_with_config`? WDTY?

sfc-gh-pjob replied:
We can't use only this method because there are commands which do not have `-c`/`--connection` option.

sfc-gh-pjob replied:
* `snowpark function init`
* `snowpark function package`

sfc-gh-turbaszek replied:
we should add this option if it's necessary for those commands

sfc-gh-pjob replied:
These commands work offline, they don't need any connection.

sfc-gh-astus replied:
We can define default connection now, this invoke will not be needed soon. Leave as it is

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        )
+
+    def assert_that_no_files_are_staged_in_snowflake(self) -> None:
+        self.assert_that_only_these_files_are_staged_in_snowflake()
```
Alias for method? 🤔

sfc-gh-pjob replied:
yes, to reflect in method name what exactly I want to assert

sfc-gh-astus commented on tests_integration/testing_utils/assertions/test_file_assertions.py:
```diff
@@ -0,0 +1,18 @@
+    assert set(os.listdir()) == set(filenames)
+
+
+def assert_that_file_content_is_equal_to_snapshot(
```
Should we care about files content?

sfc-gh-pjob replied:
IMO we should to avoid changing it by mistake (for example by replacing strings globally).

sfc-gh-astus replied:
I think this check is not needed, I can't find case when this content can change

sfc-gh-astus replied:
We copy files with external lib 

```
def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
             ignore_dangling_symlinks=False, dirs_exist_ok=False):
    """Recursively copy a directory tree and return the destination directory.
```

I think is safe :)

sfc-gh-pjob replied:
We copy files using external lib, but we keep files in our repo (we can change files content by mistake). Maybe integration tests are not the best option for such check but what is the cost of having such check here?

sfc-gh-astus commented on tests_integration/testing_utils/file_utils.py:
```diff
@@ -0,0 +1,6 @@
+def replace_text_in_file(file_path: str, to_replace: str, replacement: str) -> None:
+    with open(file_path, "r") as file:
```
Maybe just single open with `w+`?

sfc-gh-pjob replied:
I think it won't work.

* `w+` won't work because it will truncate the file (https://stackoverflow.com/a/30931305)
* `r+` also won't work because it will move the file pointer while reading
* current solution is like this one: https://stackoverflow.com/a/17141572

sfc-gh-astus commented on tests_integration/testing_utils/assertions/test_result_assertions.py:
```diff
@@ -0,0 +1,91 @@
+    )
+
+
+def assert_that_result_is_usage_error(
```
Not used function?

sfc-gh-pjob replied:
removed

sfc-gh-astus commented on tests_integration/testing_utils/assertions/test_result_assertions.py:
```diff
@@ -0,0 +1,91 @@
+    assert_that_result_is_successful(result)
+    actual = result.output.lower() if ignore_case else result.output
+    expected = expected_output.lower() if ignore_case else expected_output
+    assert expected in actual
```
Maybe mock print and assert like in other IT tests?

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        stages_in_database = [
+            stage["name"].lower() for stage in self.query_stages_in_database()
+        ]
+        if "deployments" in stages_in_database:
```
I hope other commands don't use `deployments` stage 🙏

sfc-gh-pjob replied:
I'm not sure I understand this question. What if they use the same stage?

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+            f"SHOW STAGES IN DATABASE {self.snowflake_session.database}"
+        )
+
+    def query_files_uploaded_in_this_test_case(self) -> List[Dict[str, Any]]:
```
There is easier way to do this:

```
row_from_snowflake_session(execute_single_sql("ls @deployments"))
```

This will return list of files in stage

sfc-gh-astus replied:
in dictionary

sfc-gh-pjob replied:
I don't think I understand.

1. `LS` seems to be just an alias for `LS`
2. I still have to somehow check if the stage exists (snowflake will return an error if the stage doesn't exist).
3. I still have to filter files to get only files created in this test (to avoid issues with debugging in case of bugs in other tests or in cleanups).

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        arguments="()",
+    )
+    _test_steps.assert_that_no_functions_are_in_snowflake()
+    # I'm not sure if it is expected behavior to leave staged files belonging to the removed function.
```
Will be removed after tests with database anyway

sfc-gh-pjob replied:
This my comment in the code was about the lack of removing files from a stage after removing a function. I'm removing the comment because I know now that the behaviour is correct.

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+    _test_steps.assert_that_no_functions_are_in_snowflake()
+    _test_steps.assert_that_no_files_are_staged_in_snowflake()
+
+    _test_steps.snowpark_function_list_should_return_no_data()
```
We should test it if functions are in scope of database, if yes this check is not needed

sfc-gh-pjob replied:
Can you explain what do you mean by `if functions are in scope of database`?

Current assertions in the test case check exactly the same what I check manually when testing the tool:
* empty start at the beginning (both presented by Snowflake SQL output and presented by our tool)
    * I can agree that maybe I don't have to execute SQLs here, but still IMO it's worth to check correct output of `function list`.
* successful execution of commands
* correct state of local files after commands which can change it
* correct state in Snowflake after commands which can change it (checked using SQL)
* correct presentation and usage of functions by our tool (list, describe, execute) after commands which can change functions state

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        f"{function_name}() RETURN VARCHAR"
+    )
+    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
+        f"deployments/{function_name}/app.zip"
```
Is there a way to pass stage name? If not maybe we should add it?

sfc-gh-pjob replied:
`deployments` stage is currently hardcoded (`src/snowcli/utils.py`). We can add a task for adding an option to pass custom stage name.

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        )
+        assert_that_result_is_successful_and_json_output_contains_value_at_path(
+            result=result, path=[0, "language"], expected_value="PYTHON"
+        )
```
This method can be replaced with 

```
self._setup.runner.invoke_with_config_and_integration_connection(
            [
                "--format",
                "json",
                "snowpark",
                "function",
                "list",
                "--like",
                f"{self._setup.object_name_prefix}%",
            ]
        )

    assert contains_row_with(
        row_from_mock(mock_print),
        {
            "name": function_name.upper(),
            "arguments": f"{function_name}{arguments} RETURN {result_type}".upper(),
            "language": "PYTHON", 
        },
    )
```

WDYT?

sfc-gh-pjob replied:
I don't like using mocks if we can execute an assertion on a real output. IMO it makes future refactors of printing more difficult. But as I wrote in another comment, I think we should talk about it.

sfc-gh-astus commented on tests_integration/test_snowpark_function.py:
```diff
@@ -0,0 +1,417 @@
+        function_name=function_name,
+        arguments="()",
+        result_type="NUMBER",
+    )
```
I think one test `function list` is enough

sfc-gh-pjob replied:
I described in **[this comment](https://github.com/Snowflake-Labs/snowcli/pull/245#discussion_r1273658415)** why there are few invocations of `function list`. But IMO we should talk about it. I understand it may cost us execution time.

sfc-gh-turbaszek commented on tests_integration/test_warehouse.py:
```diff
@@ -7,7 +7,7 @@
 @mock.patch("snowcli.cli.warehouse.print_db_cursor")
 def test_warehouse_status_query(mock_print, runner, snowflake_session):
-    runner.invoke_with_config(["warehouse", "status"])
+    runner.invoke_with_config_and_integration_connection(["warehouse", "status"])
```
Maybe we should introduce `integration_runner` to avoid mixing responsibilities in single fixture?

sfc-gh-pjob replied:
I'm not sure if it resolves any problem. We have two ways of execution commands in integration tests (with and without `-c integration` flag) but both are correct and both are used by integration tests.

Adding another "runner" fixture will make you think which runner to use instead of which method to use. I don't see any benefit here, only complications.

sfc-gh-pjob commented:
I rebased the branch and added two commits with fixes. Unfortunately, I did sth wrong (probably changed commit hash) and my original commit is not visible in [github commits history for this PR](https://github.com/Snowflake-Labs/snowcli/pull/245/commits) but it still visible correctly in git log (8eab8e5):
```
git log --oneline
cf57462 [SNOW-851220] Fixed invocation of CLI runner in other integration tests
b89f994 [SNOW-851220] Review fixes
8eab8e5 [SNOW-851220] Integration tests of snowpark function commands
7b26ca0 (tag: v1.0.1-rc1, origin/main, origin/HEAD, main) Bump version to 1.0.1 (#264)

```

@sfc-gh-astus , @sfc-gh-turbaszek 
Please look at my fixes and my responses to your comments.

sfc-gh-pjob commented on tests_integration/snowflake_connector.py:
```diff
@@ -12,7 +12,7 @@
 @pytest.fixture(scope="session")
 def create_database():
-    database_name = f"db_{uuid.uuid4().hex}"
+    database_name = f"snowcli_integration_tests_db_{uuid.uuid4().hex}"
```
done
